contracts/SpokePool.sol

```diff
@@ -548,59 +548,101 @@ abstract contract SpokePool is
+     * with the msg.sender address to create a unique uint256 depositNonce and ensure that the msg.sender cannot
+     * use this function to front-run another depositor's unsafe deposit. This function guarantees that the resultant
+     * deposit nonce will not collide with a safe uint256 deposit nonce whose 24 most significant bytes are always 0.
+     * @param depositor See identically named parameter in depositV3() comments.
+     * @param recipient See identically named parameter in depositV3() comments.
+     * @param inputToken See identically named parameter in depositV3() comments.
+     * @param outputToken See identically named parameter in depositV3() comments.
+     * @param inputAmount See identically named parameter in depositV3() comments.
+     * @param outputAmount See identically named parameter in depositV3() comments.
+     * @param destinationChainId See identically named parameter in depositV3() comments.
+     * @param exclusiveRelayer See identically named parameter in depositV3() comments.
+     * @param quoteTimestamp See identically named parameter in depositV3() comments.
+     * @param fillDeadline See identically named parameter in depositV3() comments.
+     * @param exclusivityPeriod See identically named parameter in depositV3() comments.
+     * @param message See identically named parameter in depositV3() comments.
+     */
+    function unsafeDepositV3(
+        address depositor,
+        address recipient,
+        address inputToken,
+        address outputToken,
+        uint256 inputAmount,
+        uint256 outputAmount,
+        uint256 destinationChainId,
+        address exclusiveRelayer,
+        uint96 depositNonce,
+        uint32 quoteTimestamp,
+        uint32 fillDeadline,
+        uint32 exclusivityPeriod,
+        bytes calldata message
+    ) public payable nonReentrant unpausedDeposits {
+        // @dev Create the uint256 deposit ID by concatenating the address (which is 20 bytes) with the 12 byte
+        // depositNonce parameter. The resultant 32 byte string can be casted to an "unsafe" uint256 deposit ID.
+        // This guarantees the resultant ID won't collide with a "safe" deposit ID which is set by
+        // implicitly casting a uint32 to a uint256, where the left-most 24 bytes are set to 0. We guarantee
+        // that there are no collisions between "unsafe" and "safe" deposit ID's as long as the msg.sender's address
+        // is not the zero address.
+
+        // @dev For some L2's, it could be possible that the zero address can be the msg.sender which might
+        // make it possible for an unsafe deposit hash to collide with a safe deposit nonce. To prevent these cases,
+        // we might want to consider explicitly checking that the msg.sender is not the zero address. However,
+        // this is unlikely and we choose to not check it and incur the extra gas cost:
+        // e.g. if (msg.sender == address(0)) revert InvalidUnsafeDepositor();
+
```

Comment:
OOC, what's the reasoning behind not wrapping all this in a hash? Simplicity for the integrator to predict the deposit hash?

Hash advantages:
- We can use full 256 bit depositNonces (might improve cross compatibility with things like permit2 or other noncing systems that may be used concurrently).
- No matter how we concat the data, there is no concern about creating collisions with normal depositIds.
- We could add more parameters that get baked into the depositId. For instance, this could be called by a contract that passes its msg.sender as the depositor. Adding the depositor to the depositId hash would mean that different depositors wouldn't collide.

Reply:
Otherwise I totally agree with all the reasons FOR hashing this

Reply:
Did the math on this.

The probability of finding 28 bytes of leading zeros is `1 - (1 - (1/2^224))^num_tries`.

You get to about 0.3% after `1e65` tries (screenshot below). For reference, Bitcoin produces about 2e28 hashes _per year_, meaning it would take about 5e36 years for the entire bitcoin network to crack to have a 0.3% chance of cracking this.


![Screenshot 2024-09-26 at 10 26 23 AM](https://github.com/user-attachments/assets/7b201adc-7f01-49bc-b069-2057c51145fc)